hbase-server/src/main/java/org/apache/hadoop/hbase/regionserver/HStore.java

```diff
@@ -1222,6 +1225,14 @@ private boolean updateStorefiles(List<HStoreFile> sfs, long snapshotId) throws I
     this.lock.writeLock().lock();
     try {
       this.storeEngine.getStoreFileManager().insertNewFiles(sfs);
+      /**
+       * NOTE:we should keep clearSnapshot method inside the write lock because clearSnapshot may
+       * close {@link DefaultMemStore#snapshot}, which may be used by
+       * {@link DefaultMemStore#getScanners}.
+       */
+      if (snapshotId > 0) {
```

Comment:
So this is the actual fix part. Moving the clearSnapshot back under the writelock. Correct?

Reply:
@anoopsjohn,  thank you for review, yes, you are right ,that is the actual fix part.

Reply:
So the problem here is that we can not call memstore.clearSnapshot and memstore.getScanners at the same time? Looking at the code, seems you are right. There is no atomic operations when getting the snapshot segment, it is possible that we put the segment into the list, and before we increase its ref count by calling getScanner on it, we could have already closed it.

Reply:
@Apache9,  thank you very much for code view, yes, you are right， and I also find that reference count implementation  of ` MemStoreLABImpl `is flawed because its checking and increasing or decreasing is not done in atomicity and it ignores state checking  when there  is illegal state in reference count(eg. increasing or decreasing when the resource is already freed) and it may hidden another bug in flushing, I would try to fix it.